src/vms/pvm/etna-builder/spend.ts

```diff
@@ -0,0 +1,473 @@
+ */
+export const splitByLocktime = (
+  utxos: readonly Utxo[],
+  minIssuanceTime: bigint,
+): { readonly locked: readonly Utxo[]; readonly unlocked: readonly Utxo[] } => {
+  const locked: Utxo[] = [];
+  const unlocked: Utxo[] = [];
+
+  for (const utxo of utxos) {
+    let utxoOwnersLocktime: bigint;
+
+    try {
+      utxoOwnersLocktime = utxo.getOutputOwners().locktime.value();
+    } catch (error) {
+      // If we can't get the locktime, we can't spend the UTXO.
+      // TODO: Is this the right thing to do?
+      // This was necessary to get tests working with testUtxos().
+      continue;
+    }
+
+    if (minIssuanceTime < utxoOwnersLocktime) {
+      locked.push(utxo);
+    } else {
+      unlocked.push(utxo);
+    }
+  }
+
+  return { locked, unlocked };
+};
+
+/**
+ * @internal
+ *
+ * Separates the provided UTXOs into two lists:
+ * - `other` contains UTXOs that have an asset ID different from `assetId`.
+ * - `requested` contains UTXOs that have an asset ID equal to `assetId`.
+ *
+ * @param utxos {readonly Utxo[]}
+ * @param assetId {string}
+ *
+ * @returns Object containing two lists of UTXOs.
+ */
+export const splitByAssetId = (
+  utxos: readonly Utxo[],
+  assetId: string,
+): { readonly other: readonly Utxo[]; readonly requested: readonly Utxo[] } => {
+  const other: Utxo[] = [];
+  const requested: Utxo[] = [];
+
+  for (const utxo of utxos) {
+    if (assetId === utxo.assetId.toString()) {
+      requested.push(utxo);
+    } else {
+      other.push(utxo);
+    }
+  }
+
+  return { other, requested };
+};
+
+/**
+ * @internal
+ *
+ * Returns the TransferOutput that was, potentially, wrapped by a stakeable lockout.
+ *
+ * If the output was stakeable and locked, the locktime is returned.
+ * Otherwise, the locktime returned will be 0n.
+ *
+ * If the output is not an error is returned.
+ */
+export const unwrapOutput = (
+  output: Serializable,
+):
+  | [error: null, transferOutput: TransferOutput, locktime: bigint]
+  | [error: Error, transferOutput: null, locktime: null] => {
+  try {
+    if (isStakeableLockOut(output) && isTransferOut(output.transferOut)) {
+      return [null, output.transferOut, output.lockTime.value()];
+    } else if (isTransferOut(output)) {
+      return [null, output, 0n];
+    }
+  } catch (error) {
+    return [
+      new Error('An unexpected error occurred while unwrapping output', {
+        cause: error instanceof Error ? error : undefined,
+      }),
+      null,
+      null,
+    ];
+  }
+
+  return [new Error('Unknown output type'), null, null];
+};
+
+type SpendResult = Readonly<{
+  changeOutputs: readonly TransferableOutput[];
+  inputs: readonly TransferableInput[];
+  inputUTXOs: readonly Utxo[];
+  stakeOutputs: readonly TransferableOutput[];
+}>;
+
+type SpendProps = Readonly<{
+  /**
+   * Contains the currently accrued transaction complexity that
+   * will be used to calculate the required fees to be burned.
+   */
+  complexity: Dimensions;
```

Comment:
This is only the initial complexity, correct? The SpendHelper will add additional complexity as it accumulates inputs and outputs?